crates/ruff/src/rules/flake8_todos/rules.rs

```diff
@@ -232,33 +232,42 @@ enum Directive {
 impl Directive {
     /// Extract a [`Directive`] from a comment.
     ///
-    /// Returns the offset of the directive within the comment, and the matching directive tag.
+    /// Returns the matching directive tag and its offset within the comment.
     fn from_comment(comment: &str) -> Option<(Directive, TextSize)> {
-        let trimmed = comment.trim_start_matches('#').trim_start();
-        let offset = comment.text_len() - trimmed.text_len();
-        let mut chars = trimmed.chars();
-        match (
-            chars.next(),
-            chars.next(),
-            chars.next(),
-            chars.next(),
-            chars.next(),
-        ) {
-            (
-                Some('F' | 'f'),
-                Some('I' | 'i'),
-                Some('X' | 'x'),
-                Some('M' | 'm'),
-                Some('E' | 'e'),
-            ) => Some((Directive::Fixme, offset)),
-            (Some('T' | 't'), Some('O' | 'o'), Some('D' | 'd'), Some('O' | 'o'), ..) => {
-                Some((Directive::Todo, offset))
-            }
-            (Some('X' | 'x'), Some('X' | 'x'), Some('X' | 'x'), ..) => {
-                Some((Directive::Xxx, offset))
+        let mut subset_opt = Some(comment);
+        let mut total_offset = TextSize::new(0);
+
+        // Loop over the comment to catch cases like `# foo # TODO`.
+        while let Some(subset) = subset_opt {
+            let trimmed = subset.trim_start_matches('#').trim_start().to_lowercase();
+
+            let offset = subset.text_len() - trimmed.text_len();
+            total_offset += offset;
+
+            let directive = if trimmed.starts_with("fixme") {
```

Comment:
Is there a more idiomatic way to do this? I couldn't think of anything

Reply:
An alternative approach could be to split the comment into sections, then iterate over those sections. We can also use a HashMap to map the directive strings to their corresponding variants, which allows us to eliminate the if else chain.

```
impl FromStr for Directive {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "fixme" => Ok(Directive::Fixme),
            "xxx" => Ok(Directive::Xxx),
            "todo" => Ok(Directive::Todo),
            _ => Err(()),
        }
    }
}

fn from_comment(comment: &str) -> Option<(Directive, TextSize)> {
    let comment_parts = comment.split('#').collect::<Vec<&str>>();

    let mut total_offset = TextSize::new(0);

    for part in comment_parts {
        let trimmed = part.trim().to_lowercase();
        total_offset += TextSize::of_str("#") + TextSize::try_from(trimmed.len()).unwrap();
        
        match trimmed.parse::<Directive>() {
            Ok(directive) => return Some((directive, total_offset)),
            Err(_) => continue,
        }
    }

    None
}
```

Reply:
I liked that. I think we could avoid collecting into a `Vec` since we're only iterating once. 

Using `spilt` is probably slightly slower because it iterates twice over the string but I wouldn't mind that in favor of readability. 

Nit: You can replace `TextSize::of_str("#")` with `'#'.text_len()` and `TextSize::try_from(trimmed.len()).unwrap()` with `trimmed.text_len()`

Reply:
I wouldn't expect the slowdown to be that significant because it only runs on comments but I'm okay either way